grpc/profiler.go

```diff
@@ -0,0 +1,245 @@
+	}
+
+	return &ProfilerClient{
+		Name:                  name,
+		serviceURL:            serviceURL,
+		ProfilerClientContext: context,
+	}, nil
+
+}
+
+// ProfilerOP is the gRPC function to provide the entry point for the profiler operations
+func (s *ProfilerServer) ProfilerOP(ctx context.Context, req *proto.ProfilerOPRequest) (*proto.ProfilerOPResponse, error) {
+	logrus.Infof("Profiler operation: %v, port: %v", req.RequestOp, req.PortNumber)
+	var err error
+	switch req.RequestOp {
+	case proto.Op_SHOW:
+		reply := &proto.ProfilerOPResponse{}
+		reply.ProfilerAddr, err = s.ShowProfiler()
+		return reply, err
+	case proto.Op_ENABLE:
+		reply := &proto.ProfilerOPResponse{}
+		reply.ProfilerAddr, err = s.EnableProfiler(req.PortNumber)
+		return reply, err
+	case proto.Op_DISABLE:
+		reply := &proto.ProfilerOPResponse{}
+		reply.ProfilerAddr, err = s.DisableProfiler()
+		return reply, err
+	default:
+		return nil, errors.New("invalid operation")
+	}
+}
+
+// ShowProfiler returns the address of the profiler
+func (s *ProfilerServer) ShowProfiler() (string, error) {
+	s.profilerLock.RLock()
+	defer s.profilerLock.RUnlock()
+
+	logrus.Info("Prepareing to show the profiler address")
+	if s.profilerServer != nil {
+		return s.profilerServer.Addr, nil
+	}
+	if s.errMsg != "" {
+		return s.errMsg, nil
+	}
+	return "", nil
+}
+
+// EnableProfiler enables the profiler with specific port number.
+// It will fail if the profiler is already enabled or the port number is invalid(0).
+// For the profiler server enabling failed, it will return the error message.
+// And keep the above error message, then user can get the error message by op `show`.
+// The normal(success) case will return the address of the profiler server.
+func (s *ProfilerServer) EnableProfiler(portNumber int32) (string, error) {
+	s.profilerLock.Lock()
+	defer s.profilerLock.Unlock()
+
+	logrus.Info("Prepareing to enable the profiler")
+
+	if s.profilerServer != nil {
+		return "", fmt.Errorf("profiler server is already running at %v", s.profilerServer.Addr)
+	}
+
+	profilerPort := int(portNumber)
+	if profilerPort == 0 {
+		s.errMsg = "enable profiler failed: invalid port number"
+		return s.errMsg, fmt.Errorf("invalid port number: %v", portNumber)
+	}
+
+	profilerAddr := fmt.Sprintf(":%d", profilerPort)
+	s.profilerServer = &http.Server{
+		Addr:              profilerAddr,
+		ReadHeaderTimeout: 10 * time.Second,
+	}
+	go func() {
+		if err := s.profilerServer.ListenAndServe(); err != http.ErrServerClosed {
+			logrus.WithError(err).Warnf("Get error when start profiler server %v", s.profilerServer.Addr)
+			s.profilerServer = nil
+			s.errMsg = err.Error()
+			return
+		}
+		logrus.Infof("Profiler server (%v) is closed", s.profilerServer.Addr)
+	}()
+
+	logrus.Infof("Waiting the profiler server(%v) to start", s.profilerServer.Addr)
+	// Wait for the profiler server to start, and check the profiler server.
+	retryCount := 3
+	for i := 0; i < retryCount; i++ {
```

Comment:
Should we consider adding delay between retries?

Reply:
Because the timeout is 1 second, if the server does not respond, we will delay 1 second to try next.
We could skip to add another delay mechanism. WDYT?